File: analysis-service/app/utils/validate.py
"""
Field validation helpers for the analysis-service.

Satisfies Requirements 3.2, 3.3, 4.1, 4.2, 4.3
"""

from app.models.schemas import RawNewsMessage


# Required fields that must be present and non-empty in a raw message dict
_REQUIRED_RAW_FIELDS = (
    "articleId",
    "sourceUrl",
    "title",
    "body",
    "sourceName",
    "publishedAt",
    "schemaVersion",
)


def validate_raw_message(data: dict) -> RawNewsMessage:
    """
    Validate a raw dict against the RawNewsMessage schema.

    Raises:
        ValueError: if any required field is missing or empty.

    Returns:
        A validated RawNewsMessage instance.
    """
    if not isinstance(data, dict):
        raise ValueError(f"Expected a dict, got {type(data).__name__}")

    missing = [field for field in _REQUIRED_RAW_FIELDS if not data.get(field)]
    if missing:
        raise ValueError(
            f"RawNewsMessage is missing or has empty required fields: {missing}"
        )

    # Delegate remaining validation (type coercion, extra-field handling) to Pydantic
    return RawNewsMessage(**data)


def validate_bias_score(score: float) -> float:
    """
    Validate that a bias score is a valid probability in [0.0, 1.0].

    Raises:
        ValueError: if score is outside the closed interval [0.0, 1.0].

    Returns:
        The validated score unchanged.
    """
    if not isinstance(score, (int, float)):
        raise ValueError(
            f"Bias score must be a numeric value, got {type(score).__name__}"
        )

    if score < 0.0 or score > 1.0:
        raise ValueError(
            f"Bias score must be in [0.0, 1.0], got {score}"
        )

    return float(score)

File: ingestion-service/src/utils/validate.ts
import { RawNewsMessage } from '../types/RawNewsMessage';

/**
 * ISO 8601 datetime regex — accepts full datetime strings with optional fractional seconds and timezone.
 * Examples: "2024-01-15T10:30:00Z", "2024-01-15T10:30:00.000Z", "2024-01-15T10:30:00+05:30"
 */
const ISO_8601_REGEX =
  /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?(Z|[+-]\d{2}:\d{2})$/;

/**
 * Validates an unknown input and returns a typed RawNewsMessage if all required fields are present
 * and well-formed. Throws a descriptive Error if any required field is missing or invalid.
 *
 * Required fields: title, sourceName (source), publishedAt (timestamp), articleId, sourceUrl,
 * body, schemaVersion.
 *
 * Satisfies Requirements 1.4, 2.1, 2.2, 2.3
 *
 * @param input - The raw unknown value to validate (typically parsed JSON from an external source)
 * @returns A fully-typed RawNewsMessage
 * @throws Error with a descriptive message if validation fails
 */
export function validateRawArticle(input: unknown): RawNewsMessage {
  if (input === null || typeof input !== 'object') {
    throw new Error('Validation failed: input must be a non-null object');
  }

  const record = input as Record<string, unknown>;

  // --- title ---
  if (!record['title'] || typeof record['title'] !== 'string' || record['title'].trim() === '') {
    throw new Error(
      'Validation failed: "title" is required and must be a non-empty string'
    );
  }

  // --- sourceName (source) ---
  if (
    !record['sourceName'] ||
    typeof record['sourceName'] !== 'string' ||
    record['sourceName'].trim() === ''
  ) {
    throw new Error(
      'Validation failed: "sourceName" is required and must be a non-empty string'
    );
  }

  // --- publishedAt (timestamp) ---
  if (
    !record['publishedAt'] ||
    typeof record['publishedAt'] !== 'string' ||
    record['publishedAt'].trim() === ''
  ) {
    throw new Error(
      'Validation failed: "publishedAt" is required and must be a non-empty string'
    );
  }

  if (!ISO_8601_REGEX.test(record['publishedAt'].trim())) {
    throw new Error(
      `Validation failed: "publishedAt" must be a valid ISO 8601 datetime string (received: "${record['publishedAt']}")`
    );
  }

  // --- articleId ---
  if (
    !record['articleId'] ||
    typeof record['articleId'] !== 'string' ||
    record['articleId'].trim() === ''
  ) {
    throw new Error(
      'Validation failed: "articleId" is required and must be a non-empty string'
    );
  }

  // --- sourceUrl ---
  if (
    !record['sourceUrl'] ||
    typeof record['sourceUrl'] !== 'string' ||
    record['sourceUrl'].trim() === ''
  ) {
    throw new Error(
      'Validation failed: "sourceUrl" is required and must be a non-empty string'
    );
  }

  // --- body ---
  if (typeof record['body'] !== 'string') {
    throw new Error(
      'Validation failed: "body" is required and must be a string'
    );
  }

  // --- schemaVersion ---
  if (
    !record['schemaVersion'] ||
    typeof record['schemaVersion'] !== 'string' ||
    record['schemaVersion'].trim() === ''
  ) {
    throw new Error(
      'Validation failed: "schemaVersion" is required and must be a non-empty string'
    );
  }

  return {
    articleId: record['articleId'].trim(),
    sourceUrl: record['sourceUrl'].trim(),
    title: record['title'].trim(),
    body: record['body'] as string,
    sourceName: record['sourceName'].trim(),
    publishedAt: record['publishedAt'].trim(),
    schemaVersion: record['schemaVersion'].trim(),
  };
}
